ocaml/xapi/xapi_session.ml

```diff
@@ -661,6 +666,62 @@ let login_no_password_common ~__context ~uname ~originator ~host ~pool
   ignore (Client.Pool.get_all ~rpc ~session_id) ;
   session_id
 
+let login_no_password_common ~__context ~uname ~originator ~host ~pool
+    ~is_local_superuser ~subject ~auth_user_sid ~auth_user_name
+    ~rbac_permissions ~db_ref ~client_certificate =
+  Context.with_tracing ~originator ~__context __FUNCTION__ @@ fun __context ->
+  let is_useable_session session_id =
+    let is_valid_session session_id =
+      try
+        (* Call an API function to check the session is still valid *)
+        let rpc = Helpers.make_rpc ~__context in
+        ignore (Client.Pool.get_all ~rpc ~session_id) ;
+        true
+      with Api_errors.Server_error (err, _) ->
+        debug "%s: Invalid session: %s" __FUNCTION__ err ;
+        false
+    in
+    session_id <> Ref.null
+    && ((not !Xapi_globs.validate_reusable_pool_session)
+       || is_valid_session session_id
+       )
+  in
+  let create_session () =
+    let new_session_id =
+      login_no_password_common_create_session ~__context ~uname ~originator
+        ~host ~pool ~is_local_superuser ~subject ~auth_user_sid ~auth_user_name
+        ~rbac_permissions ~db_ref ~client_certificate
+    in
+    new_session_id
+  in
+  let rec get_session () =
+    let session = Atomic.get reusable_pool_session in
+    if is_useable_session session then (
+      (* Check if the session changed during validation.
+         Use our version regardless to avoid being stuck in a loop of session creation *)
+      if Atomic.get reusable_pool_session <> session then
+        debug
+          "reusable_pool_session has changed, using the original version anyway" ;
+      session
+    ) else
+      let new_session = create_session () in
+      if Atomic.compare_and_set reusable_pool_session Ref.null new_session then
+        new_session
+      else (
```

Comment:
The `reusable_pool_session` may be an expired (not `Ref.null`) one, so the `is_usable_session` above returns `false` and the `Atomic.compare_and_set` returns `false`. So it would reach this `else` for loop forever?

Reply:
Great point, thank you! If !Xapi_globs.validate_reusable_pool_session is true then yes we could get into that situation. I've changed the Atomic.compare_and_set to compare against the original session we get instead of Ref.null (so it could be Ref.null or as you say it could be an expired session)